trafficFlow/carFollowingModel/lanes/baseLane.py

```diff
@@ -21,16 +24,38 @@ class BaseLane:
         place the vehicles in a default manner on the lane
     """
 
-    def __init__(self):
-        self.number_of_vehicles = 0
+    def __init__(self, full_length=100):
+        self.full_length = full_length
+
         self.road = None
         self.vehicles = []
+        self.lane_right = None
+        self.lane_left = None
         self.initialized = False
 
     def add_vehicle(self, vehicle):
         self.vehicles.append(vehicle)
         vehicle.lane = self
-        self.number_of_vehicles = self.number_of_vehicles + 1
+
+    def get_number_of_vehicles(self):
+        return len(self.vehicles)
+
+    def get_speed_difference(self, vehicle1, vehicle2):
+        if not vehicle2:
+            return math.inf
+        if not vehicle1:
+            return -math.inf
+        return vehicle2.velocity - vehicle1.velocity
+
+    def lane_changes(self):
+        for vehicle in self.vehicles:
+            vehicle.lane_change()
```

Comment:
This may be the cause of your problem. ``vehicle.lane_change`` might add or remove elements from ``self.vehicles`` while you are iterating over it. You should copy ``self.vehicles`` before iterating over it ([or use another solution](https://stackoverflow.com/questions/3752618/python-adding-element-to-list-while-iterating) if copying is too slow.

Reply:
That's a good point, I will have a look at this.